File: iOS/Code/Xamarin/bugTrap/bugTrapKit/Trackers/Common/Tracker.cs
using System;
using System.Threading.Tasks;
using System.Collections.Generic;

namespace bugTrapKit
{
	public class Tracker
	{
		public TrackerType TrackerType { get; private set; }

		public BugDetails Details { get; } = new BugDetails ();

		public Keychain<TrackerType> Keychain { get; }

		public bool Authenticated { get; set; }

		public ITrackerState TrackerState { get; set; }

		public ITrackerProxy TrackerProxy { get; set; }

		// public Dictionary<int, BugDetails.DetailType> TrackerDetails { get; set; }

		public bool HasAuthCredentials {
			get { return TrackerType.Valid(Keychain.GetStoredKeyValues()); }
		}

		public Tracker (TrackerType trackerType)
		{
			TrackerType = trackerType;
			// Details = new BugDetails ();
			Keychain = new Keychain<TrackerType> (trackerType);
			// TrackerDetails = new Dictionary<int, BugDetails.DetailType> ();
		}

		public async Task<bool> Authenticate (Dictionary<DataKeys, string> trackerData)
		{
			if (HasAuthCredentials) {

				try {
					
					var authResult = await TrackerProxy.Authenticate(trackerData);
				
					Authenticated = authResult;
				
					await PostAuthDataFetch();

				} catch (Exception ex) {

					Authenticated = false;

					Console.WriteLine(ex.Message);
				} 
			}

			return Authenticated;
		}

		public virtual Task PostAuthDataFetch ()
		{
			throw new NotImplementedException ();
		}

		public virtual Task<bool> LoadDataForActiveDetail ()
		{
			return Task.FromResult(true);
		}

		public virtual bool UpdateIssueFromSelection (int selectedIndex)
		{
			return false;
		}

		public virtual Task<bool> CreateIssue ()
		{
			return Task.FromResult(true);
		}
	}
}

File: iOS/Code/Swift/bugTrap/bugTrapKit/Trackers/Common/Tracker.swift
//
//  Trackers.swift
//  bugTrap
//
//  Created by Colby L Williams on 8/23/14.
//  Copyright (c) 2014 bugTrap. All rights reserved.
//

import Foundation

class Tracker {
	
	let type: TrackerType
	
	let details: BugDetails
    
	
	
	let keychain: Keychain<TrackerType>
	
	var Authenticated = false
	
	var hasAuthCredentials: Bool {
		return type.valid(keychain.GetStoredKeyValues())
	}
	
	
	func authenticate(data trackerData: [DataKeys : String], callback: (Result<Bool>) -> ()) {
		
		if hasAuthCredentials {
			trackerProxy.authenticate(data: trackerData, callback: { authResult in
				switch authResult {
				case let .Error(error):
					Log.error(self.type.string, error)
					self.Authenticated = false
					callback(.Error(error))
				case let .Value(result):
					if let success = result.value {
						self.Authenticated = success
						callback(.Value(Wrapped(true)))
						self.postAuthDataFetch()
					} else {
						self.Authenticated = false
						callback(.Error(NSError(domain: self.type.string, code: 401, userInfo: nil)))
					}
				}
			})
		} else {
			self.Authenticated = false
			callback(.Error(NSError(domain: type.string, code: 401, userInfo: nil)))
		}
	}
	
	func postAuthDataFetch() { }
	
    var trackerState: TrackerState!
    
    var trackerProxy: TrackerProxy!
    
    var TrackerDetails: Dictionary<Int, BugDetails.DetailType>

    init(trackerType: TrackerType) {
		self.type = trackerType
		self.details = BugDetails()
        self.keychain = Keychain<TrackerType>(service: trackerType)
        self.TrackerDetails = Dictionary<Int, BugDetails.DetailType>()// [Int : BugDetails.DetailType]()
	}
	
	
	
    func loadDataForActiveDetail(callback: (Result<Bool>) -> ()) { }
    
    func updateIssueFromSelection(selectedIndex: Int) -> Bool {
        return false
    }
    
    func createIssue(callback: (Result<Bool>) -> ()) { }
}